Allow :if and :unless options

aliismayilov opened the pull request:
We've inherited a project which has `strip_attributes` enabled for all the ActiveRecord classes with no option to exclude certain classes.

By having `:if` and/or `:unless` options, we could achieve a cleaner solution to exclude certain tables from stripping.

```ruby
class ApplicationRecord < ActiveRecord::Base
  strip_attributes if: :strip_attributes?

  def strip_attributes?
    true
  end
end
```

Then we can override `strip_attributes?` in child classes.

rmm5t commented:
Thank you for the effort and though that you put into this, but it's best [not to push your pull requests](https://www.igvita.com/2011/12/19/dont-push-your-pull-requests/) like this without first opening an issue. 

1. I would not have advised that the project ever apply `strip_attributes` like that globally. There are dozens of ways to configure strip_attributes per model, and performing the defaults globally probably isn't a great choice.  However, I don't know how similar all your models and fields are, so ¯\\\_(ツ)\_/¯.

2. There's no need to modify strip_attributes to do what you want. Ruby has support for this kind of behavior built in already and without having to patch this gem. Here's a quick (untested) example off the top of my head:

```ruby
class ApplicationRecord < ActiveRecord::Base
  def self.inherited(subclass)
    subclass.strip_attributes if subclass.strip_attributes?
  end

  def self.strip_attributes?
    true
  end
end

class ModelWithoutStripAttributes < ApplicationRecord
  def self.strip_attributes?
    false
  end
end
```

aliismayilov commented:
Thank you for the tip! The `self.inherited` won't work, because it's invoked at the moment of `class ModelWithoutStripAttributes < ApplicationRecord`, which means that it'll pick its parent's version of `self.strip_attributes?`.

I'm still curious, what's a good way of letting the child class to override parent's strip_attributes line.

rmm5t commented:
Okay, Let me noodle on this some more. Reopening this PR, but no promises.

rmm5t commented on lib/strip_attributes.rb:
```diff
@@ -38,6 +38,14 @@ module StripAttributes
+
+    if options && options[:unless] && record_or_string.send(options[:unless].to_sym)
+      return record_or_string
+    end
```
[PLEASE MOVE] These two guard clauses belong as a part of the `strip_record` method, not here. This method accepts a record or a string. Calling `send` on a passed string doesn't make sense here. 

[PLEASE ADD] Also, Rails conventions allow for a `Proc` to be passed to `:if` and `:unless` options in addition to a symbol representing an instance method. Basically, if the value is a symbol or string, use `send`, but if the value `respond_to?(:call)` and the `arity == 1`, use `call` passing in the record. This would allow for something like:

```ruby
strip_attributes if: Proc.new { |user| user.signup_step > 2 }
```

rmm5t commented on test/strip_attributes_test.rb:
```diff
@@ -242,6 +262,54 @@ def test_should_strip_unicode
+    assert_equal "fiz \n  fiz", record.fiz
+    assert_equal "",            record.baz
+    assert_equal " ",           record.bang
+  end
```
Nice tests!

aliismayilov commented:
I understand that it wouldn't really make sense to skip strip_attributes per instance in real world, but it'd make sense to do it per class. I also did my part of research to find out what's a nicer way of overriding before_validation on a child class (Stackoverflow, hallelujah!).

Regarding the Rails API allowing to use if/unless as procs, I've simply delegated it to rails. I think it looks cleaner now.

rmm5t commented on lib/strip_attributes.rb:
```diff
@@ -5,7 +5,12 @@ module ActiveModel::Validations::HelperMethods
+      unless: options && options[:unless]
+    }
+
+    before_validation(rails_opts) do |record|
```
[KUDOS] I love, love, love 😍this as an implementation decision. It make me way more comfortable about this feature and this PR. Great idea to just pass the options to `before_validation`. I hadn't thought of that, but it makes total sense.

[PLEASE CHANGE] However, I have a code cleanup request before merging. Please get rid of the `rails_opts` variable and change this to:

```ruby
before_validation(options.slice(:if, :unless)) do |record|
  StripAttributes.strip(record, options)
end
```

aliismayilov commented:
glad that makes you happy too!

I tried to get rid of `rails_opts` variable, I ended up with the following line:
```ruby
before_validation(options && options.slice(:if, :unless) || {}) do |record|
```

because `before_validation` isn't cool with receiving `nil` as its options. To be honest, I did not like the above line very much. So I've added another commit, where I changed the default of options from `nil` to an empty hash. Please let me know which approach is you think is better. I can as well discard the last commit.

rmm5t commented:
Good work. I like it. I don’t remember why I chose to make options default to nil in the first place instead of any empty hash.  

I just gave it all a quick review in mobile. I’ll give it a final review and merge once I’m on a bigger keyboard. 

A versioned release will come soon after that. 

Thank you for your persistence here, even against my early stubbornness. I like the final outcome.

rmm5t commented:
v1.9.0 is now out in the wild!
https://rubygems.org/gems/strip_attributes
